README.rst
========
bringbuf
========

*bringbuf* is very basic circular buffer implementation. Its purpose is to handle byte streams (for instance from a serial port). It is based on enque, which is a efficient way to handle queues in Python. First bytes written to the buffer, are first bytes read from the buffer. If the buffer is full it overflows and overwrites the oldest data.

The code is available at `github`_.

.. _github: https://github.com/ticktronaut/bringbuf

Install
-------

Requirements
````````````

There are no requirements. *bringbuf* has view dependencies (*collections*, *itertools* and *warnings*), which are core modules.

Setup
`````

Install *bringbuf* using the following command:

::

    $ [sudo] pip install bringbuf 

Usage
-----

.. code-block:: python

    #!/usr/bin/env python
    #coding=utf-8
    
    # The following example shows the basic usage of bRingBuf class
    
    from bringbuf import bRingBuf
    
    # define a ring buffer with the size of 5 bytes
    buf = bRingBuf(5)
    
    # the buffers length initially is zero
    print(buf.len)
    
    # check if buffer is empty (True)
    print(buf.is_empty())
    
    # enqueue 6 bytes to the buffer
    # (note that the buffer will overflow)
    buf.enqueue([0x01, 0x02, 0x03, 0x04, 0x05, 0x06])
    
    # now five bytes are in the buffer (maximal number of bytes)
    print(buf.len)
    
    # check if buffer is empty (False) 
    print(buf.is_empty())
    
    # read content of the whole buffer without removing any bytes
    print(buf.read(buf.len))
    
    # read and remove three bytes from the buffer 
    print(buf.dequeue(3))
    
    # read content of the whole buffer without removing any bytes
    print(buf.read(buf.len))
    
    # enque four new bytes
    buf.enqueue([0x07, 0x08, 0x09])
    
    # read content of the whole buffer without removing any bytes
    print(buf.read(buf.len))
    
    # check whether buffer contains a pattern?
    # this might be solved by buf.index, which also returns the index,
    # but this is a more descriptive way to check
    print(buf.contains([0x02, 0x08]),  buf.contains([0x07, 0x08, 0x09]))
    
    # index of patterns
    print(buf.index([0x02, 0x08]), buf.index([0x07, 0x08, 0x09]))
    
    # contains pattern (with offset)?
    print(buf.contains([0x02, 0x08], 2), buf.contains([0x07, 0x08, 0x09], offset=2))
    
    # index of patterns (with offset)
    print(buf.index([0x02, 0x08], 2), buf.index([0x07, 0x08, 0x09], 2))
    
    # read with offset
    print(buf.read(3, 3))
    
    # enqueue a single byte
    buf.enqueue_byte(0xff)
    
    print(buf.read(buf.len))
    
    # dequeue a single byte
    print(buf.dequeue_byte())

License
-------

Copyright (c) 2016 Andreas Gschossmann

Permission is hereby granted, free of charge, to any person obtaining a copy of
this software and associated documentation files (the "Software"), to deal in
the Software without restriction, including without limitation the rights to
use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
of the Software, and to permit persons to whom the Software is furnished to do
so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


CHANGES.txt
v0.0.dev1, 16.03.2018 -- Initial release.


bringbuf/bringbuf.py
#!/usr/bin/env python
#coding=utf-8

from collections import deque
import itertools
import warnings

class bRingBuf(object):
    """bRingBuf is a simple circular buffer to handle byte streams.
 
    It is based on enqueue, which is a efficient way to handle queues in
    Python. The main purpose of this library is to handle bytestreams, 
    like for instance from a serial port.

    Example
    -------
    ::
        from bringbuf import bRingBuf
        
        buf = bRingBuf(5)
        print(buf.is_empty())
        buf.enqueue([0x01, 0x02, 0x03, 0x04, 0x05, 0x06])
        print(buf.is_empty())
        print(buf.read(buf.len))
        print(buf.dequeue(3))
        print(buf.read(buf.len))

    Attributes
    ----------
    _max_size : int
        Maximal number that the ring buffer can carry, before overlfow.
    _b : bytearray
        Private byte container variable. It may be removed in future versions.
    _q : deque
        Python queue object, containing the buffer data.
    len : int
        Number of bytes written to the buffer.
    
    """
    def __init__(self, max_size=100):
        self._max_size = max_size
        #self._b = bytearray(max_size) # bytes vs. bytearray http://ze.phyr.us/bytearray/
        self._q = deque(maxlen=max_size)
        self.len = 0

    def enqueue(self, b):
        """Write bytes to the buffer.

        Parameters
        ----------
        b
            Iterable array of bytes to enqueue to the buffer

        """
        self._q.extend(b)
        self.len = len(self._q)
         

    def enqueue_byte(self, b):
        """Write a single byte to the buffer.

        Parameters
        ----------
        b
            byte to enqueue to the buffer

        """
        self._q.append(b)
        self.len = len(self._q)

    def dequeue(self, n=1):
        """Read and remove number of bytes from the buffer.

        Parameters
        ----------
        n
            Number of byes to read and remove from the buffer (default is one).
        return
            Bytes read and removed from buffer.
        """
        if n > self.len: 
            message = str(n) + ' bytes requested, but only ' + str(self.len) + ' bytes available.'
            n = self.len 
            warnings.warn(message)
        b = bytearray(n)
        for i in range(n):
            b[i] = self._q.popleft()

        self.len = len(self._q)
        return b

    def dequeue_byte(self):
        """Read and remove a single byte from the buffer.

        Parameters
        ----------
        return
            single byte read and removed from buffer.
        """
        b = self._q.popleft()

        self.len = len(self._q)
        return b

    def read(self, n=1, offset=0):
        """Read bytes in the buffer, without removing them from the buffer.

        Parameters
        ----------
        n
            Number of bytes to read from the buffer.
        offset
            Index offset to start reading bytes.
        return
            Bytes read from buffer.
        """
        if ((n+offset) > self.len):
            message = str(n) + ' bytes requested, but only ' + str( (self.len-offset) ) + ' bytes available.'
            n = self.len - offset
            warnings.warn(message)
        # todo: return bytes type by using bytes (...), which is an immutable type
        #return list( itertools.islice( self._q,offset,( n+offset ) ) )
        return bytes( itertools.islice( self._q,offset,( n+offset ) ) )

    def clear(self):
        """Clear all bytes from buffer.
        """
        self.len = 0
        self._q.clear()

    def is_empty(self):
        """Returns true if buffer is empty.

        Parameters
        ----------
        return
            Return is True if buffer is empty, else False.
        """
        return not bool(self.len)

    def contains(self, pattern, offset=0):
        """Returns true if b is available in queue.
        
        Parameters
        ----------
        pattern 
            Pattern to search for (iterable byte sequence).
        offset
            Index offset to start.
        return
            Return True if buffer contains pattern, else False.
        """
        if pattern[0] in self.read((self.len-offset), offset):
            index = ( self.read(self.len, offset).index(pattern[0]) ) + offset
            return ( list(self.read(len(pattern), index)) == pattern )
        else:
            return False 

    def index(self, pattern, offset=0):
        """If the buffer includes the pattern, this function returns its index. Otherwise it return -1

        Parameters
        ----------
        pattern
            Pattern to search for (iterable byte sequence).
        offset
            Index offset to start.
        return
            Index of occurence of pattern. If buffer does not contain pattern return value is -1.
        """
        if self.contains(pattern, offset):
            return self.read( (self.len-offset), offset ).index(pattern[0]) + offset
        else:
            return -1


examples/sample.py
#!/usr/bin/env python
#coding=utf-8

# The following example shows the basic usage of bRingBuf class

from bringbuf import bRingBuf

# define a ring buffer with the size of 5 bytes
buf = bRingBuf(5)

# the buffers length initially is zero
print(buf.len)

# check if buffer is empty (True)
print(buf.is_empty())

# enqueue 6 bytes to the buffer
# (note that the buffer will overflow)
buf.enqueue([0x01, 0x02, 0x03, 0x04, 0x05, 0x06])

# now five bytes are in the buffer (maximal number of bytes)
print(buf.len)

# check if buffer is empty (False) 
print(buf.is_empty())

# read content of the whole buffer without removing any bytes
print(buf.read(buf.len))

# read and remove three bytes from the buffer 
print(buf.dequeue(3))

# read content of the whole buffer without removing any bytes
print(buf.read(buf.len))

# enque four new bytes
buf.enqueue([0x07, 0x08, 0x09])

# read content of the whole buffer without removing any bytes
print(buf.read(buf.len))

# check whether buffer contains a pattern?
# this might be solved by buf.index, which also returns the index,
# but this is a more descriptive way to check
print(buf.contains([0x02, 0x08]),  buf.contains([0x07, 0x08, 0x09]))

# index of patterns
print(buf.index([0x02, 0x08]), buf.index([0x07, 0x08, 0x09]))

# contains pattern (with offset)?
print(buf.contains([0x02, 0x08], 2), buf.contains([0x07, 0x08, 0x09], offset=2))

# index of patterns (with offset)
print(buf.index([0x02, 0x08], 2), buf.index([0x07, 0x08, 0x09], 2))

# read with offset
print(buf.read(3, 3))

# enqueue a single byte
buf.enqueue_byte(0xff)

print(buf.read(buf.len))

# dequeue a single byte
print(buf.dequeue_byte())
